File: src/lib/evidence.ts
import { ExtractedCase } from "./types";
import {
  ingestSource,
  findReusableSource,
  retrieveChunks,
  fetchCourtListenerSource,
  fetchWikipediaSource,
  deleteExpiredSources,
} from "./rag";
import type { FetchedSource, RetrievedChunk } from "./rag";

export interface Evidence {
  caseInfo?: {
    caseName: string | null;
    defendant: string | null;
    victim: string | null;
    crimeType: string | null;
    jurisdiction: string | null;
    state: string | null;
    approximateYear: string | null;
  };

  originalText?: string;

  wikipedia?: {
    title: string;
    text: string;
    url: string;
  };

  courtlistener?: {
    title: string;
    text: string;
    url: string;
    court?: string;
    dateFiled?: string;
  };

  /**
   * RAG-retrieved chunks from the shared knowledge base.
   * Populated during Evidence Assembly when top search results are
   * ingested and retrieved. Additive and non-fatal: may be empty if
   * RAG is unavailable or fails.
   */
  ragChunks?: RetrievedChunk[];
}

/**
 * The #1 CourtListener and #1 Wikipedia search results, preserved from
 * the search stage so that Evidence Assembly can use them as RAG sources
 * and as concise narrative/contextual evidence.
 *
 * CourtListener results are primarily RAG ingestion candidates (legal
 * source corpus for retrieval). Wikipedia results provide concise
 * narrative/case context when available.
 */
export interface SearchContext {
  courtlistener?: {
    title: string;
    url: string;
    snippet: string;
    court?: string;
    dateFiled?: string;
    /** CourtListener cluster id from the search result; used to resolve the
     *  underlying opinion record(s) via the Opinions API. */
    clusterId?: string;
  };
  wikipedia?: {
    title: string;
    url: string;
    summary: string;
  };
}

// Limit long text/transcript extractions to first 60% and last 40% of maxChars
function limitText(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text;

  const head = Math.floor(maxChars * 0.6);
  const tail = Math.floor(maxChars * 0.4);

  return text.slice(0, head) + "\n\n[Middle omitted]\n\n" + text.slice(-tail);
}

/**
 * Build a retrieval query string from the extracted case signals.
 * Used as the query text for RAG chunk retrieval.
 */
function buildRagQuery(extracted: ExtractedCase): string {
  const parts = [
    extracted.caseName,
    extracted.defendant,
    extracted.victim,
    extracted.crimeType,
    extracted.jurisdiction,
    extracted.state,
    extracted.approximateYear,
    ...(extracted.keywords ?? []),
  ].filter(Boolean);

  return parts.join(" ");
}

/**
 * Ingest an external source document into the RAG knowledge base, reusing
 * an existing unexpired source when present.
 *
 * The cheap DB check runs FIRST so that expensive external fetching,
 * chunking, and embedding are skipped when a valid source already exists.
 * This is especially important for CourtListener, whose Clusters/Opinions
 * endpoints are token-authenticated and rate-limited.
 */
async function ingestExternalSource(
  label: string,
  sourceUrl: string,
  fetchSource: () => Promise<FetchedSource | null>,
  extractedMeta: Record<string, unknown>,
): Promise<void> {
  const existing = await findReusableSource(sourceUrl);

  if (existing) {
    console.log(
      `RAG: ${label} source reused — ${sourceUrl} ` +
        `(${existing.chunkCount} chunks)`,
    );
    return;
  }

  // Only hit the external API when the source is not already cached.
  const source = await fetchSource();

  if (!source) {
    console.error(
      `RAG: ${label} source unavailable for ${sourceUrl}; skipping`,
    );
    return;
  }

  try {
    const result = await ingestSource({
      url: source.url,
      sourceType: source.sourceType,
      title: source.title,
      sourceText: source.sourceText,
      extractedMeta,
    });

    console.log(
      `RAG: ${label} ${source.url} — ` +
        `${result.reused ? "reused existing source" : "ingested new source"}; ` +
        `${result.chunkCount} chunks`,
    );
  } catch (err) {
    console.error(`RAG: ${label} ingest failed for ${source.url}:`, err);
  }
}

// Fetch evidence from the original source, search results, and RAG.
export async function fetchEvidence(
  extracted: ExtractedCase,
  originalText: string,
  searchContext?: SearchContext,
): Promise<Evidence> {
  // Always include original extracted input
  const evidence: Evidence = {
    caseInfo: {
      caseName: extracted.caseName,
      defendant: extracted.defendant,
      victim: extracted.victim,
      crimeType: extracted.crimeType,
      jurisdiction: extracted.jurisdiction,
      state: extracted.state,
      approximateYear: extracted.approximateYear,
    },
    originalText: limitText(originalText, 14000),
  };

  console.log("Evidence caseInfo:", JSON.stringify(evidence.caseInfo, null, 2));

  // Wikipedia: use the concise summary as narrative/contextual evidence
  // when a Wikipedia search result exists.
  const wiki = searchContext?.wikipedia;
  if (wiki?.title) {
    evidence.wikipedia = {
      title: wiki.title,
      text: limitText(wiki.summary, 6000),
      url: wiki.url,
    };
  }

  // CourtListener: the top search result's snippet is preserved as a
  // lightweight evidence layer. The full opinion is ingested into RAG
  // below and retrieved as relevant chunks.
  const court = searchContext?.courtlistener;
  if (court?.title) {
    evidence.courtlistener = {
      title: court.title,
      text: court.snippet ?? "",
      url: court.url,
      court: court.court,
      dateFiled: court.dateFiled,
    };
  }

  // RAG: ingest the full underlying documents of the top search results and
  // retrieve relevant chunks. This is additive and non-fatal.
  //
  // Before ingestion, remove any expired sources. This allows the existing
  // cleanup/TTL behavior to determine whether a source needs to be re-ingested.
  try {
    await deleteExpiredSources();

    const extractedMeta = {
      caseName: extracted.caseName,
      defendant: extracted.defendant,
      victim: extracted.victim,
      crimeType: extracted.crimeType,
      jurisdiction: extracted.jurisdiction,
      state: extracted.state,
      approximateYear: extracted.approximateYear,
      keywords: extracted.keywords,
      confidence: extracted.confidence,
    };

    // Top CourtListener result (opinion)
    if (court?.clusterId) {
      const clusterId = court.clusterId;
      await ingestExternalSource(
        "CourtListener",
        court.url,
        () => fetchCourtListenerSource(clusterId),
        extractedMeta,
      );
    }

    // Top Wikipedia result (full article)
    if (wiki?.title) {
      const title = wiki.title;
      const url = wiki.url;
      await ingestExternalSource(
        "Wikipedia",
        url,
        () => fetchWikipediaSource(title, url),
        extractedMeta,
      );
    }

    // Retrieve relevant chunks
    const queryText = buildRagQuery(extracted);

    if (queryText) {
      const chunks = await retrieveChunks(queryText, {
        topK: 3,
      });

      evidence.ragChunks = chunks;

      let ragChunksCount = 0;

      console.log(`RAG: retrieved ${chunks.length} chunks`);

      evidence.ragChunks.forEach((chunk, index) => {
        ragChunksCount += chunk.text.length;

        console.log(`===== RAG CHUNK ${index + 1} =====`);
        console.log(chunk);
        console.log(`===== END RAG CHUNK ${index + 1} =====`);
      });

      // Evidence-size diagnostics.
      console.log("Evidence sizes:");

      console.log(
        "Original text:",
        evidence.originalText?.length ?? 0,
        "chars",
        "≈",
        Math.ceil((evidence.originalText?.length ?? 0) / 4),
        "tokens",
      );

      console.log(
        "Wikipedia:",
        evidence.wikipedia?.text.length ?? 0,
        "chars",
        "≈",
        Math.ceil((evidence.wikipedia?.text.length ?? 0) / 4),
        "tokens",
      );

      console.log(
        "CourtListener:",
        evidence.courtlistener?.text.length ?? 0,
        "chars",
        "≈",
        Math.ceil((evidence.courtlistener?.text.length ?? 0) / 4),
        "tokens",
      );

      console.log(
        "RAG chunks:",
        ragChunksCount,
        "chars",
        "≈",
        Math.ceil(ragChunksCount / 4),
        "tokens",
      );

      const serializedEvidence = JSON.stringify(evidence);

      console.log(
        "Total evidence:",
        serializedEvidence.length,
        "chars",
        "≈",
        Math.ceil(serializedEvidence.length / 4),
        "tokens",
      );
    }
  } catch (err) {
    console.error("RAG retrieval failed:", err);
  }

  return evidence;
}

File: python/casefile/evidence/evidence.py
"""Evidence Assembly — ports src/lib/evidence.ts.

Combines extracted case info, original text, Wikipedia summary, CourtListener
snippet, and RAG-retrieved chunks into an Evidence object for the overview
generation stage. Runs active RAG cleanup, reuses unexpired sources, and is
additive + non-fatal if RAG fails.
"""

import json
import logging
from typing import Any, Optional

from ..rag import (
    delete_expired_sources,
    fetch_courtlistener_source,
    fetch_wikipedia_source,
    find_reusable_source,
    ingest_source,
    retrieve_chunks,
)
from ..rag.types import IngestInput, RetrievedChunk
from ..types import ExtractedCase

logger = logging.getLogger(__name__)


class Evidence:
    """The evidence object passed to overview generation."""

    def __init__(self) -> None:
        self.caseInfo: Optional[dict[str, str | None]] = None
        self.originalText: Optional[str] = None
        self.wikipedia: Optional[dict[str, str]] = None
        self.courtlistener: Optional[dict[str, str | None]] = None
        self.ragChunks: list[RetrievedChunk] = []

    def to_dict(self) -> dict[str, Any]:
        """Serialize to a dict for JSON prompt embedding."""
        return {
            "caseInfo": self.caseInfo,
            "originalText": self.originalText,
            "wikipedia": self.wikipedia,
            "courtlistener": self.courtlistener,
            "ragChunks": [c.model_dump() for c in self.ragChunks],
        }


class SearchContext:
    """Preserved #1 search results from the search stage (for RAG ingestion)."""

    def __init__(self) -> None:
        self.courtlistener: Optional[dict[str, str | None]] = None
        self.wikipedia: Optional[dict[str, str | None]] = None


def _limit_text(text: str, max_chars: int) -> str:
    """Limit text to first 60% + last 40% of maxChars (mirrors evidence.ts)."""
    if len(text) <= max_chars:
        return text

    head = int(max_chars * 0.6)
    tail = int(max_chars * 0.4)

    return text[:head] + "\n\n[Middle omitted]\n\n" + text[-tail:]


def _build_rag_query(extracted: ExtractedCase) -> str:
    """Build a retrieval query string from extracted case signals."""
    parts = [
        extracted.caseName,
        extracted.defendant,
        extracted.victim,
        extracted.crimeType,
        extracted.jurisdiction,
        extracted.state,
        extracted.approximateYear,
        *(extracted.keywords or []),
    ]
    return " ".join(p for p in parts if p)


def _ingest_external_source(
    label: str,
    source_url: str,
    fetch_fn,
    extracted_meta: dict[str, Any],
) -> None:
    """Ingest an external source, reusing existing unexpired sources."""
    existing = find_reusable_source(source_url)

    if existing:
        logger.info(
            "RAG: %s source reused - %s (%d chunks)",
            label,
            source_url,
            existing["chunkCount"],
        )
        return

    # Only hit the external API when the source is not already cached.
    source = fetch_fn()

    if not source:
        logger.error("RAG: %s source unavailable for %s; skipping", label, source_url)
        return

    try:
        result = ingest_source(
            IngestInput(
                url=source.url,
                sourceType=source.sourceType,
                title=source.title,
                sourceText=source.sourceText,
                extractedMeta=extracted_meta,
            )
        )
        logger.info(
            "RAG: %s %s - %s; %d chunks",
            label,
            source.url,
            "reused existing source" if result.reused else "ingested new source",
            result.chunkCount,
        )
    except Exception as err:
        logger.error("RAG: %s ingest failed for %s: %s", label, source.url, err)


def fetch_evidence(
    extracted: ExtractedCase,
    original_text: str,
    search_context: Optional[SearchContext] = None,
) -> Evidence:
    """Assemble evidence from source text, search results, and RAG."""
    evidence = Evidence()
    evidence.caseInfo = {
        "caseName": extracted.caseName,
        "defendant": extracted.defendant,
        "victim": extracted.victim,
        "crimeType": extracted.crimeType,
        "jurisdiction": extracted.jurisdiction,
        "state": extracted.state,
        "approximateYear": extracted.approximateYear,
    }
    evidence.originalText = _limit_text(original_text, 14000)

    logger.info("Evidence caseInfo: %s", json.dumps(evidence.caseInfo, indent=2))

    # Wikipedia: use the concise summary as narrative/contextual evidence.
    wiki = search_context.wikipedia if search_context else None
    if wiki and wiki.get("title"):
        evidence.wikipedia = {
            "title": wiki["title"],
            "text": _limit_text(wiki.get("summary") or "", 6000),
            "url": wiki.get("url") or "",
        }

    # CourtListener: preserve the top search result's snippet.
    court = search_context.courtlistener if search_context else None
    if court and court.get("title"):
        evidence.courtlistener = {
            "title": court["title"],
            "text": court.get("snippet") or "",
            "url": court.get("url") or "",
            "court": court.get("court"),
            "dateFiled": court.get("dateFiled"),
        }

    # RAG: ingest the full underlying documents of the top search results and
    # retrieve relevant chunks. Additive and non-fatal.
    try:
        delete_expired_sources()

        extracted_meta = {
            "caseName": extracted.caseName,
            "defendant": extracted.defendant,
            "victim": extracted.victim,
            "crimeType": extracted.crimeType,
            "jurisdiction": extracted.jurisdiction,
            "state": extracted.state,
            "approximateYear": extracted.approximateYear,
            "keywords": extracted.keywords,
            "confidence": extracted.confidence,
        }

        # Top CourtListener result (opinion).
        if court and court.get("clusterId"):
            cluster_id = court["clusterId"]
            _ingest_external_source(
                "CourtListener",
                court.get("url") or "",
                lambda ci=cluster_id: fetch_courtlistener_source(ci),
                extracted_meta,
            )

        # Top Wikipedia result (full article).
        if wiki and wiki.get("title"):
            _ingest_external_source(
                "Wikipedia",
                wiki.get("url") or "",
                lambda t=wiki["title"], u=wiki.get("url") or "": fetch_wikipedia_source(t, u),
                extracted_meta,
            )

        # Retrieve relevant chunks.
        query_text = _build_rag_query(extracted)

        if query_text:
            chunks = retrieve_chunks(query_text, top_k=3)
            evidence.ragChunks = chunks

            logger.info("RAG: retrieved %d chunks", len(chunks))
            for index, chunk in enumerate(chunks, start=1):
                logger.info("===== RAG CHUNK %d =====", index)
                logger.info(chunk.model_dump())
                logger.info("===== END RAG CHUNK %d =====", index)
    except Exception as err:
        logger.error("RAG retrieval failed: %s", err)

    return evidence